[BREAKING] sync combine with select (#2158)
JuliaData/DataFrames.jl: docs/src/man/split_apply_combine.md

Many data analysis tasks involve splitting a data set into groups, applying some functions to each of the groups and then combining the results. A standardized framework for handling this sort of computation is described in the paper "[The Split-Apply-Combine Strategy for Data Analysis](http://www.jstatsoft.org/v40/i01)", written by Hadley Wickham.

The DataFrames package supports the split-apply-combine strategy through the `by` function, which is a shorthand for `groupby` followed by `map` and/or `combine`. `by` takes in three arguments: (1) a `DataFrame`, (2) one or more columns to split the `DataFrame` on, and (3) a specification of one or more functions to apply to each subset of the `DataFrame`. This specification can be of the following forms:
1. a `col => function` pair indicating that `function` should be called with the vector
   of values for column `col`, which can be a column name or index
2. a `cols => function` pair indicating that `function` should be called with

>    positional arguments holding columns `cols`, which can be a any valid column selector

Reviewer: Maybe merge these two cases? The distinction was needed before as a named tuple was passed for 2, but now 1 is a special case of 2.

Author: OK - done.

> 3. a `col => function => target_col` `cols => function => target_col` form additionally

Reviewer: ```suggestion
3. a `cols => function => target_col` form additionally
```

   specifying the name of the target column (this assumes that `function` returns a single value or a vector)
4. a `col => target_col` pair, which renames the column `col` to `target_col`

> 5. a `nrow` or `nrow => target_col` form which efficiently produces number of rows in a group

Reviewer: ```suggestion
5. a `nrow` or `nrow => target_col` form which efficiently computes the number of rows in a group
```

>    (without `target_col` a new column created gets `:nrow` name)

Reviewer: ```suggestion
   (without `target_col` the new column is called `:nrow`)
```

> 6. several arguments of the forms given above or vectors of pairs for these forms

Reviewer: ```suggestion
6. several arguments of the forms given above, or vectors thereof
```

7. a function which will be called with a `SubDataFrame` corresponding to each group;

>    this form should be avoided due to its poor performance unless a `SubDataFrame`
>    has a very large number of columns (in which case `SubDataFrame` avoids excessive
>    compilation)

Reviewer: ```suggestion
   this form should be avoided due to its poor performance unless a very large
   number of columns are processed (in which case `SubDataFrame` avoids excessive
   compilation)
```

> Forms 1 to 5 and 7 can be also passed as a first argument to `by`.

Reviewer: ```suggestion
All forms except 6 can be also passed as the first argument to `by`.
```

> In all of these cases, the `function` can return either a single row or multiple rows.
> Normally it is assumed that a function returns a single value or a vector.
> Additionally multiple columns in a form of `AbstractDataFrame`, `AbstactMatrix`,
> `NamedTuple` or `DataFrameRow` can be returned by the `function` if `by` is passed
> exactly one `function` and `target_col` is not specified.

Reviewer: ```suggestion
In all of these cases, `function` can return either a single row or multiple rows.
`function` can always generate a single column by returning a single value or a vector.
Additionally, if `by` is passed exactly one `function` and `target_col` is not specified,
`function` can return multiple columns in the form of an `AbstractDataFrame`,
`AbstractMatrix`, `NamedTuple` or `DataFrameRow`.
```

julia> by(iris, :Species,
          [:PetalLength, :SepalLength] =>

>           (p, s) -> (a=mean(p)/mean(s), b=sum(p))) # multiple arguments are splatted

Reviewer: There's no splatting from the user POV here.
```suggestion
          (p, s) -> (a=mean(p)/mean(s), b=sum(p))) # multiple columns are passed as arguments
```

julia> for (key, subdf) in pairs(groupby(iris, :Species))
           println("Number of data points for $(key.Species): $(nrow(subdf))")
       end
Number of data points for Iris-setosa: 50
Number of data points for Iris-versicolor: 50
Number of data points for Iris-virginica: 50
```

The value of `key` in the previous example is a [`DataFrames.GroupKey`](@ref) object,
which can be used in a similar fashion to a `NamedTuple`.

> Grouping a data frame using `groupby` function can be seen as adding a lookup key

Reviewer: ```suggestion
Grouping a data frame using `groupby` the function can be seen as adding a lookup key
```

> to it. Doing such lookups is efficient and can be also performed by passing
> a `Tuple` or `NamedTuple` directly:

Reviewer: Why "directly"?
```suggestion
to it. Such lookups can be performed efficiently by indexing the resulting
`GroupedDataFrame` with a a `Tuple` or `NamedTuple`:
```
